### PR TITLE
CHECK_PREFER_IS_NOT: Exempt predicative method calls

### LINKED ISSUE #362: false positives: Prefer IS NOT to NOT IS
Hello, 

this check creates false positives for code like
`    IF NOT is_loaded( )`.
where is_loaded( ) has an ABAP_BOOL returning parameter; or, for a more complex example,
`    IF is_new( ) AND NOT is_loaded( )`.

I don't think the code would be better readable if I changed it to 
`    IF is_loaded( ) = abap_false.`
`    IF is_new( ) =abap_true AND is_loaded( ) = abap_false.`
or is that what Clean ABAP recommends in such a case?

Similarly, the check creates messages for code like 
 `   IF NOT ( a < b AND c < d OR a > b AND c > d ).`

Of course, you could replace this with
`    IF ( ( a >= b OR c >= d ) AND ( a <= b OR c <= d ) ).`
but usually when if saw the IF NOT ( … ) syntax, there was a good reason for it, and the logical expression was easier to understand that way. 

Anyway, such cases do not quite fit to the error message "Prefer IS NOT to NOT IS". Is this check simply searching the code for "IF NOT"? Maybe it could at least restrict the findings to cases where an actual "IS" appears somewhere in the logical expression …

Kind regards, 
Jörg-Michael

### PR DESCRIPTION
**Pull request by ConjuringCoffee**
Fixes #362

### CODE REVIEW

**Comment by ConjuringCoffee** [comment]
I based the implementation of `ltc_predicative_method` on the previous test classes like `ltc_contains_any_of`. However, should the test classes really all inherit from `ltc_not_value`? That seems wrong. I'm going to correct my new test class and the test classes that come before...

File: src/checks/y_check_prefer_is_not.clas.abap [hunk 1]
```diff
@@ -74,5 +82,8 @@ CLASS y_check_prefer_is_not IMPLEMENTATION.
                       OR token-str CP 'MATCHES*' ).
   ENDMETHOD.
 
+  METHOD is_predicative_method.
+    result = xsdbool( token-str CP '*(' ).
+  ENDMETHOD.
```
Comment (bjoern-jueliger-sap) [inline src/checks/y_check_prefer_is_not.clas.abap:87]: Unfortunately, this is too simple and also recognizes non-predicative calls, as in:
```abap
report test.

class main definition.
  public section.
    class-methods get_number
      returning value(number) type i.
endclass.

CLASS main IMPLEMENTATION.

  METHOD get_number.
  ENDMETHOD.

ENDCLASS.

start-of-selection.
  if not main=>get_number( ) = 1. "<---there should be a finding here
  endif.
```

**Review by bjoern-jueliger-sap — CHANGES_REQUESTED** [review]

File: src/checks/y_check_prefer_is_not.clas.abap (hunk 1 above)
Comment (ConjuringCoffee) [reply src/checks/y_check_prefer_is_not.clas.abap:87]: Oh no, I forgot about that... let me see if I can fix that.

Comment (ConjuringCoffee) [reply src/checks/y_check_prefer_is_not.clas.abap]: Can you think of other cases that aren't covered by the reworked logic?

Comment (ConjuringCoffee) [reply src/checks/y_check_prefer_is_not.clas.abap]: Well, now I understand why nobody attempted to solve this issue in the past two and a half years. 😅 Thank you for your help!

**Comment by ConjuringCoffee** [comment]
Over the weekend I found another case that wasn't covered previously: `IF NEW lcl_object( )->is_active( ).`

**Review by bjoern-jueliger-sap — APPROVED** [review]
Looks good to me and works fine on all my weird examples now - thanks for your effort!

**Pull request merged**